table/block_based/block_based_table_builder.cc

```diff
@@ -1886,10 +1899,76 @@ void BlockBasedTableBuilder::EnterUnbuffered() {
                                         r->pending_handle);
       }
     }
-
+    // TODO: figure out when we can ensure all the buffered data blocks are
+    // flushed and call this function
+    // TODO: figure out what is the right status message for successfully
+    // flushing all the buffered data blocks
+    if (ok()) {
+      FreeAllBufferedDataBlocksMemFromCache();
```

Comment:
This appears to be inside the loop over `r->data_block_buffers` so it'll execute a lot of times.

Also we need to find a way to call this even when there's an error (might be called in multiple places).

Reply:
> Also we need to make sure to call this in the error case so a failed table building releases its dummy entry handles.

I changed the code to release memory only after the line `r->data_block_buffers.clear();` and removed my previous buggy `ok()` check since the memory reserved should stay as close as the total bytes in `r->data_block_buffers` (which is `r->data_begin_offset` I believe) regardless of any failure that can cause a non-okay status up to this point. Not sure if this is what your comment was suggesting :)